src/coreclr/vm/comdelegate.cpp

```diff
@@ -1941,7 +1908,68 @@ PCODE COMDelegate::GetInvokeMethodStub(EEImplMethodDesc* pMD)
         if (*pMD->GetSig() != (IMAGE_CEE_CS_CALLCONV_HASTHIS | IMAGE_CEE_CS_CALLCONV_DEFAULT))
             COMPlusThrow(kInvalidProgramException);
 
-        ret = COMDelegate::TheDelegateInvokeStub();
+        Stub *pStub = pClass->m_pMultiCastInvokeStub;
+        if (pStub == NULL)
+        {
+            MetaSig sig(pMD);
+
+            BOOL fReturnVal = !sig.IsReturnTypeVoid();
+
+            SigTypeContext emptyContext;
+            ILStubLinker sl(pMD->GetModule(), pMD->GetSignature(), &emptyContext, pMD, (ILStubLinkerFlags)(ILSTUB_LINKER_FLAG_STUB_HAS_THIS | ILSTUB_LINKER_FLAG_TARGET_HAS_THIS));
+
+            ILCodeStream *pCode = sl.NewCodeStream(ILStubLinker::kDispatch);
+
+            // Store the function pointer into local variable to avoid unnecessary register usage by JIT
+            DWORD numMethodPtr = pCode->NewLocal(ELEMENT_TYPE_I);
+            
+            // Load the method pointer
+            pCode->EmitLoadThis();
+            pCode->EmitLDFLD(pCode->GetToken(CoreLibBinder::GetField(FIELD__DELEGATE__METHOD_PTR)));
+            pCode->EmitSTLOC(numMethodPtr);
+
+            // Load the target object
+            pCode->EmitLoadThis();
+            pCode->EmitLDFLD(pCode->GetToken(CoreLibBinder::GetField(FIELD__DELEGATE__TARGET)));
+
+            // Load the arguments
+            for (UINT paramCount = 0; paramCount < sig.NumFixedArgs(); paramCount++)
+                pCode->EmitLDARG(paramCount);
+
+            PCCOR_SIGNATURE pSig;
+            DWORD cbSig;
+            pMD->GetSig(&pSig,&cbSig);
+
+            // call the delegate
+            pCode->EmitLDLOC(numMethodPtr);
+            pCode->EmitCALLI(pCode->GetSigToken(pSig, cbSig), sig.NumFixedArgs() + 1, fReturnVal); // NumFixedArgs doesn't include "this"
```

Comment:
This should be `pCode->EmitCALL(pCode->GetToken(pMD), sig.NumFixedArgs(), fReturnVal);` like in other delegate stubs. It will fix the test failures, and it will also make the KeepAlive unnecessary (once the JIT is fixed).

Reply:
Do you mean making the stub recursive like intrinsic, to let JIT kick-in? What about NativeAOT in this case?

Reply:
I'm really not fully understanding the underlying issue. `_methodPtr` should follow the same calling convention with delegate `Invoke` method, and it would point to a stub to handle calling convention difference or static methods etc. `_methodPtrAux` is the original method entry, and not directly invoked.

In other words, why can it fail if doing the same thing with JIT expanded invocation?

Reply:
The custom calling convention is necessary for delegates that point to virtually dispatched methods. The delegate invocation callsite must use specific registers so that the virtual dispatch can do its part. For example, https://github.com/dotnet/runtime/blob/851779a2360c63a30494376d5f169eff6d1783f1/src/coreclr/vm/i386/virtualcallstubcpu.hpp#L650-L670 talks about this.

> Do you mean making the stub recursive like intrinsic, to let JIT kick-in? 

Yes, I think it should end up looking like the recursive intrinsic.

> What about NativeAOT in this case?

NativeAOT does not use the custom calling convention for delegate invocation. Instead, it has a special stub to deal with the virtually dispatched delegates. It is less efficient, but simpler.

Reply:
OK I see. It's another special constraint that can't be expressed in IL. Sometimes JIT may choose to put the function pointer in another register.
But I'm still not 100% understanding that JIT expanded code puts delegate in RAX with `jmp [rax+0x18]`, while the asm stub loads the function pointer with `mov rax, [rax+0x18]` then `jmp rax`. The register usages are not totally the same.

Reply:
Using recursive call doesn't fix the test failure. A local run asserts with `IS_ALIGNED(dest, sizeof(SIZE_T))` in `InlinedMem
moveGCRefsHelper`. It may be caused by invalid assumption causing something else treated as GC reference. I'll debug it later.

Reply:
It doesn't fix the test failure. I'm consistently seeing assert of `IS_ALIGNED(dest, sizeof(SIZE_T))` in `BulkMoveWithWriteBarrier`. Looks like some invalid address treated as GC reference. I'll debug it further.

Reply:
It seems that my local build was corrupt. After a full rebuild I'm able to reproduce the CI failure with `dotnet test`. However, extracting the failing test into Main runs successfully. Since this PR is touching delegate and unit test runner uses it extensively, the issue may be caused by mixture usages of different delegates. I'll try to trace more.

Reply:
It's end up a type to misuse multicast stub as the single cast stub. It's triggered only when a delegate was used as multicast, then single cast indirectly.
DI uses dynamic methods more and exposes this issue.